Typescript 4.8 support

### Issue workflow progress

_Progress of the issue based on the [Contributor Workflow](https://github.com/the-guild-org/Stack/blob/master/CONTRIBUTING.md#a-typical-contributor-workflow)_

- [x] 1. The issue provides a [minimal reproduction](https://en.wikipedia.org/wiki/Minimal_reproducible_example) available on [Stackblitz](https://stackblitz.com/fork/node).
  - _Please install the latest `@envelop/*` packages that you are using._
  - _Please make sure the reproduction is as small as possible._
- [ ] 2. A failing test has been provided
- [ ] 3. A local solution has been provided
- [ ] 4. A pull request is pending review

---

**Describe the bug**

Project's using ts@>4.8.* will not compile when skipLibCheck is disabled:

```
node_modules/@envelop/core/typings/enable-if.d.ts:13:93 - error TS2344: Type 'PluginContextType' does not satisfy the constraint 'Record<string, any>'.

13 export declare function enableIf<PluginContextType = {}>(condition: boolean, plugin: Plugin<PluginContextType> | (() => Plugin<PluginContextType>)): PluginOrDisabledPlugin;
                                                                                               ~~~~~~~~~~~~~~~~~

  node_modules/@envelop/core/typings/enable-if.d.ts:13:34
    13 export declare function enableIf<PluginContextType = {}>(condition: boolean, plugin: Plugin<PluginContextType> | (() => Plugin<PluginContextType>)): PluginOrDisabledPlugin;
                                        ~~~~~~~~~~~~~~~~~~~~~~
    This type parameter might need an `extends Record<string, any>` constraint.

node_modules/@envelop/core/typings/enable-if.d.ts:13:128 - error TS2344: Type 'PluginContextType' does not satisfy the constraint 'Record<string, any>'.

13 export declare function enableIf<PluginContextType = {}>(condition: boolean, plugin: Plugin<PluginContextType> | (() => Plugin<PluginContextType>)): PluginOrDisabledPlugin;
                                                                                                                                  ~~~~~~~~~~~~~~~~~

  node_modules/@envelop/core/typings/enable-if.d.ts:13:34
    13 export declare function enableIf<PluginContextType = {}>(condition: boolean, plugin: Plugin<PluginContextType> | (() => Plugin<PluginContextType>)): PluginOrDisabledPlugin;
                                        ~~~~~~~~~~~~~~~~~~~~~~
    This type parameter might need an `extends Record<string, any>` constraint.

node_modules/@envelop/core/typings/plugins/use-error-handler.d.ts:4:91 - error TS2344: Type 'ContextType' does not satisfy the constraint 'Record<string, any>'.

4 export declare const useErrorHandler: <ContextType>(errorHandler: ErrorHandler) => Plugin<ContextType>;
                                                                                            ~~~~~~~~~~~

  node_modules/@envelop/core/typings/plugins/use-error-handler.d.ts:4:40
    4 export declare const useErrorHandler: <ContextType>(errorHandler: ErrorHandler) => Plugin<ContextType>;
                                             ~~~~~~~~~~~
    This type parameter might need an `extends Record<string, any>` constraint.
```

**To Reproduce**
Steps to reproduce the behavior:

install typescript 4.8 in a project using envelop
set `skipLibCheck` to false
run a typescript build

repro: https://github.com/hayes/envelop-ts-4.8-bug

<!-- A clear and concise description of what you expected to happen. -->

**Environment:**

- OS:
- NodeJS:
- `@envelop/*` versions:
  - `@envelop/core`: `???`

**Additional context**

<!-- Add any other context about the problem here. -->

--- a/package.json
+++ b/package.json
@@ -77,7 +77,7 @@
     "rimraf": "3.0.2",
     "ts-jest": "27.1.3",
     "ts-node": "10.7.0",
-    "typescript": "4.7.4",
+    "typescript": "4.8.2",
     "wait-on": "6.0.1"
   },
   "resolutions": {

--- a/packages/core/src/enable-if.ts
+++ b/packages/core/src/enable-if.ts
@@ -16,7 +16,7 @@ export function isPluginEnabled(t: PluginOrDisabledPlugin): t is Plugin {
 /**
  * Utility function to enable a plugin.
  */
-export function enableIf<PluginContextType = {}>(
+export function enableIf<PluginContextType extends Record<any, any> = {}>(
   condition: boolean,
   plugin: Plugin<PluginContextType> | (() => Plugin<PluginContextType>)
 ): PluginOrDisabledPlugin {

--- a/packages/core/src/plugins/use-error-handler.ts
+++ b/packages/core/src/plugins/use-error-handler.ts
@@ -10,14 +10,16 @@ type ErrorHandlerCallback<ContextType> = {
 };
 
 const makeHandleResult =
-  <ContextType>(errorHandler: ErrorHandler) =>
+  <ContextType extends Record<any, any>>(errorHandler: ErrorHandler) =>
   ({ result, args }: ErrorHandlerCallback<ContextType>) => {
     if (result.errors?.length) {
       errorHandler(result.errors, args);
     }
   };
 
-export const useErrorHandler = <ContextType>(errorHandler: ErrorHandler): Plugin<ContextType> => {
+export const useErrorHandler = <ContextType extends Record<string, any>>(
+  errorHandler: ErrorHandler
+): Plugin<ContextType> => {
   const handleResult = makeHandleResult<ContextType>(errorHandler);
   return {
     onExecute() {

--- a/packages/plugins/execute-subscription-event/src/index.ts
+++ b/packages/plugins/execute-subscription-event/src/index.ts
@@ -18,7 +18,7 @@ export type ContextFactoryType<TContextValue = DefaultContext> = (
   options: ContextFactoryOptions
 ) => PromiseOrValue<ContextFactoryHook<TContextValue> | void>;
 
-export const useExtendContextValuePerExecuteSubscriptionEvent = <TContextValue = unknown>(
+export const useExtendContextValuePerExecuteSubscriptionEvent = <TContextValue extends Record<any, any>>(
   createContext: ContextFactoryType<TContextValue>
 ): Plugin<TContextValue> => {
   return {

--- a/packages/plugins/generic-auth/src/index.ts
+++ b/packages/plugins/generic-auth/src/index.ts
@@ -126,7 +126,7 @@ export function defaultProtectSingleValidateFn<UserType>(
 
 export const useGenericAuth = <
   UserType extends {} = {},
-  ContextType = DefaultContext,
+  ContextType extends Record<any, any> = DefaultContext,
   CurrentUserKey extends string = 'currentUser'
 >(
   options: GenericAuthPluginOptions<UserType, ContextType, CurrentUserKey>
@@ -212,7 +212,7 @@ export const useGenericAuth = <
         const user = await options.resolveUserFn(context as unknown as ContextType);
         extendContext({
           [contextFieldName]: user,
-        } as unknown as ContextType);
+        } as any);
       },
     };
   }
@@ -223,7 +223,7 @@ export const useGenericAuth = <
 
         extendContext({
           [contextFieldName]: user,
-        } as unknown as ContextType);
+        } as any);
       },
     };
   }

